### PR TITLE
Completed first pass preneeds

### PR DESCRIPTION
**Pull request by mphprogrammer**
BASED OFF OF MASTER
------------------------
- Gets a list of all cemeteries
- Gets a list of all discharge types
- Gets a list of all attachment types
- Gets a list of all states
- Gets a list of all branches of service
- Gets a list of all ranks for a branch of service

### CODE REVIEW

**Comment by ayaleloehr** [comment]
Similar to what I wrote in https://github.com/department-of-veterans-affairs/vets-api/pull/1034#issuecomment-307080324 recently about file names, this naming also worries me. Is there a more descriptive name we can give these files/classes to be sure that it's clear what form this is for, especially given that there is already another "burials" form going on Vets.gov this month?

**Comment by mphprogrammer** [comment]
@ayaleloehr Was going to call it `preneeds`, but realized that the code would nearly be exact with the "needs" burial application when someone dies. Settled for generic "burial" since that would cover all cases, but if its too generic I can name it something more acceptable. I'll ask for opinions tomorrow.

**Comment by saneshark** [comment]
@ayaleloehr to your point, I think that is a broader vets-engineering conversation. I agree with you and I think perhaps we should start name-spacing models and controller accordingly.

**Comment by mphprogrammer** [comment]
@ayaleloehr @saneshark I've changed names to something less vague.

File: app/controllers/v0/branches_of_service_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module V0
+  class BranchesOfServiceController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/branches_of_service_controller.rb]: why not put this in the parent controller so it is inherited behavior?

File: app/controllers/v0/cemeteries_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module V0
+  class CemeteriesController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/cemeteries_controller.rb]: same comment

File: app/controllers/v0/discharge_types_controller.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module V0
+  class DischargeTypesController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/discharge_types_controller.rb]: same comment

File: app/controllers/v0/military_ranks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module V0
+  class MilitaryRanksController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/military_ranks_controller.rb]: same comment

File: app/controllers/v0/preneeds_attachment_types_controller.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module V0
+  class PreneedsAttachmentTypesController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/preneeds_attachment_types_controller.rb]: same comment

File: app/controllers/v0/preneeds_states_controller.rb [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module V0
+  class PreneedsStatesController < PreneedsController
+    # We call authenticate_token because auth is optional on this endpoint.
+    skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/preneeds_states_controller.rb]: same comment

File: app/controllers/v0/branches_of_service_controller.rb (hunk 1 above)
Comment (saneshark) [inline app/controllers/v0/branches_of_service_controller.rb]: additionally can we move this and all other controllers into its own namespace / module? We're adding 5-6 new controllers and I think it will clean things up a bit if we namespace.

File: app/models/branches_of_service.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+# Cemetery model
+class BranchesOfService < Common::Base
+  include ActiveModel::Validations
+
+  # Cemetery numbers are 3-digits, implying < 1000 in total - we want all of them for populating application forms
+  per_page 1000
+  max_per_page 1000
+
+  validates :code, :begin_date, :end_date, :flat_full_descr,
+            :full_descr, :short_descr, :state_required, :upright_full_descr, presence: true
```
Comment (saneshark) [inline app/models/branches_of_service.rb]: Do we need all of these validations? As I understand it, Cemetery is a get request, so we don't need to validate the GET request. However, if you're also using this same object for building the form, I can understand the need for validation, but doesn't the application just require sending a cemetery code? Why all of the additional validation?

File: app/models/cemetery.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# frozen_string_literal: true
+require 'common/models/base'
+
+# Cemetery model
+class Cemetery < Common::Base
+  include ActiveModel::Validations
+
+  validates :cemetery_type, inclusion: { in: %w(S N P I A M) }
+  validates :name, :num, presence: true
```
Comment (saneshark) [inline app/models/cemetery.rb]: Same comment about validation. I want to make sure we're only validating cemetery based on user input, not based on what is provided to us from EOAS

File: app/models/discharge_type.rb [hunk 9]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+require 'common/models/base'
+
+# DischargeType model
+class DischargeType < Common::Base
+  include ActiveModel::Validations
+
+  validates :description, :id, presence: true
```
Comment (saneshark) [inline app/models/discharge_type.rb]: Is the discharge description user input?

File: app/models/military_rank.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+# frozen_string_literal: true
+require 'common/models/base'
+
+# Military Rank model
+class MilitaryRank < Common::Base
+  include ActiveModel::Validations
+
+  validates :branch_of_service_cd, :military_rank_detail, presence: true
```
Comment (saneshark) [inline app/models/military_rank.rb]: same comment about only validating user input

File: app/models/military_rank_detail.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+require 'common/models/base'
+
+# Military Rank details
+class MilitaryRankDetail < Common::Base
+  include ActiveModel::Validations
+
+  validates :branch_of_service_code, :rank_code, :rank_descr, presence: true
```
Comment (saneshark) [inline app/models/military_rank_detail.rb]: same comment

File: app/models/preneeds_attachment_type.rb [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+require 'common/models/base'
+
+# DischargeType model
+class PreneedsAttachmentType < Common::Base
+  include ActiveModel::Validations
+
+  validates :description, :attachment_type_id, presence: true
```
Comment (saneshark) [inline app/models/preneeds_attachment_type.rb]: same comment

Comment (saneshark) [inline app/models/preneeds_attachment_type.rb]: for attachments I think we will likely need additional validations such as individual file size, total file size of all attachments, maximum number of attachments, but perhaps those validations would go on the attachments aggregate (somewhere on the form object itself).

File: app/models/preneeds_state.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+require 'common/models/base'
+
+# PreneedsState model
+class PreneedsState < Common::Base
+  include ActiveModel::Validations
+
+  validates :code, :first_five_zip, :last_five_zip, :lower_indicator, :name, presence: true
```
Comment (saneshark) [inline app/models/preneeds_state.rb]: if this is part of a GET request only I don't think we need to validate it.

File: config/preneeds/xsd/preneeds.xsd [hunk 14]
```diff
@@ -0,0 +1,1094 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Published by JAX-WS RI at http://jax-ws.dev.java.net. RI's version is
```
Comment (saneshark) [inline config/preneeds/xsd/preneeds.xsd]: Why is this file needed in vets-api? Is it used somehow that I'm not seeing?

File: config/routes.rb [hunk 15]
```diff
@@ -100,6 +100,16 @@
...
 
+    scope :preneeds do
+      resources :cemeteries, only: :index, defaults: { format: :json }
+      resources :preneeds_states, only: :index, defaults: { format: :json }
+      resources :preneeds_attachment_types, only: :index, defaults: { format: :json }
+      resources :discharge_types, only: :index, defaults: { format: :json }
+      resources :military_ranks, only: :index, defaults: { format: :json }
+      resources :branches_of_service, only: :index, defaults: { format: :json }
+      resources :applications, only: :create, defaults: { format: :json }
+    end
```
Comment (saneshark) [inline config/routes.rb]: when you move all of these controllers into a directory, you will likely need to define the namespace as well.

File: lib/preneeds/middleware/response/preneeds_parser.rb [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        def military_rank_for_branch_of_service_list
+          return nil unless @parsed_json.keys.include?(:military_rank_for_branch_of_service) &&
+                            @parsed_json[:military_rank_for_branch_of_service].present?
+
+          @parsed_json[:military_rank_for_branch_of_service].each do |r|
+            r[:military_rank_detail] = r.delete(:id)
+          end
+
+          @parsed_json[:military_rank_for_branch_of_service]
+        end
```
Comment (saneshark) [inline lib/preneeds/middleware/response/preneeds_parser.rb]: Is it really necessary to convert XML to JSON just to de-serialize into objects? Is this how the other SOAP based clients are doing it? I would think XML could be used to marshall objects just as well, but I could be wrong.

File: lib/preneeds/middleware/response/soap_to_json.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+          dump = Hash.from_xml Ox.dump(env.body)
+
+          endpoint_resource = dump['Envelope']['Body'].keys[0]
+          returns = dump['Envelope']['Body'][endpoint_resource]['return']
+
+          env.response_headers['content-type'] = env.response_headers['content-type'].gsub('xml', 'json')
+          env.body = { endpoint_resource&.gsub(/Response|get|receive/, '')&.underscore&.downcase => returns }
+        end
+      end
```
Comment (saneshark) [inline lib/preneeds/middleware/response/soap_to_json.rb]: I thought maybe we already had some middlewares for handling SOAP

File: app/models/branches_of_service.rb [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    code
+  end
+
+  def begin_date=(value)
+    super(Time.parse(value).utc.strftime('%Y-%m-%d')) if value.is_a?(String)
+  end
+
+  def end_date=(value)
+    super(Time.parse(value).utc.strftime('%Y-%m-%d')) if value.is_a?(String)
+  end
```
Comment (saneshark) [inline app/models/branches_of_service.rb]: Letse use Virtus Coercers for this rather than override the setter methods. You can see Common::UTCTime as an example.

File: app/controllers/v0/preneeds/military_ranks_controller.rb [hunk 19]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module V0
+  module Preneeds
+    class MilitaryRanksController < PreneedsController
+      # We call authenticate_token because auth is optional on this endpoint.
+      skip_before_action(:authenticate)
```
Comment (saneshark) [inline app/controllers/v0/preneeds/military_ranks_controller.rb]: missed one.

File: app/models/preneeds/branches_of_service.rb [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+      code
+    end
+
+    def begin_date=(value)
+      super(Time.parse(value).utc.strftime('%Y-%m-%d')) if value.is_a?(String)
+    end
+
+    def end_date=(value)
+      super(Time.parse(value).utc.strftime('%Y-%m-%d')) if value.is_a?(String)
+    end
```
Comment (saneshark) [inline app/models/preneeds/branches_of_service.rb]: can we use a Coercer for these instead? See Common::UTCTime as an example.

File: config/preneeds/xsd/preneeds.xsd (hunk 14 above)
Comment (saneshark) [inline config/preneeds/xsd/preneeds.xsd]: lets remove this file.

File: config/settings.yml [hunk 21]
```diff
@@ -119,6 +119,14 @@ appeals_status:
   app_token: PUBLICDEMO123
   host: https://dsva-appeals-certification-dev-1895622301.us-gov-west-1.elb.amazonaws.com
 
+# Settings for (preneeds) burials.
+# TODO: enpoint needs to be set up for https and production api when available
+preneeds:
+  endpoint: http://qitcalpha03.cem.va.gov:8000/eoas_SOA/PreNeedApplicationPort
+  preneeds_wsdl: 'config/preneeds/wsdl/preneeds.wsdl'
+  preneeds_xsd: 'config/preneeds/xsd/preneeds.xsd'
+  schemas: 'lib/preneeds/schemas/'
```
Comment (saneshark) [inline config/settings.yml]: why is this needed? also we should use vets-json-schema

File: config/settings.yml [hunk 22]
```diff
@@ -119,6 +119,14 @@ appeals_status:
   app_token: PUBLICDEMO123
   host: https://dsva-appeals-certification-dev-1895622301.us-gov-west-1.elb.amazonaws.com
 
+# Settings for (preneeds) burials.
+# TODO: enpoint needs to be set up for https and production api when available
+preneeds:
+  endpoint: http://qitcalpha03.cem.va.gov:8000/eoas_SOA/PreNeedApplicationPort
+  preneeds_wsdl: 'config/preneeds/wsdl/preneeds.wsdl'
+  preneeds_xsd: 'config/preneeds/xsd/preneeds.xsd'
```
Comment (saneshark) [inline config/settings.yml]: I don't think this file is needed at all, please remove file and this entry.

File: config/settings.yml [hunk 23]
```diff
@@ -119,6 +119,14 @@ appeals_status:
   app_token: PUBLICDEMO123
   host: https://dsva-appeals-certification-dev-1895622301.us-gov-west-1.elb.amazonaws.com
 
+# Settings for (preneeds) burials.
+# TODO: enpoint needs to be set up for https and production api when available
+preneeds:
+  endpoint: http://qitcalpha03.cem.va.gov:8000/eoas_SOA/PreNeedApplicationPort
```
Comment (saneshark) [inline config/settings.yml]: i'm not sure we actually want to commit this URL in settings.yml -- anyone needing to use this should add these to their local versions of this file.

File: lib/preneeds/configuration.rb [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+    def connection
+      @faraday ||= Faraday.new(base_path, headers: base_request_headers, request: request_options) do |conn|
+        conn.options.open_timeout = Settings.preneeds.open_timeout
+        conn.options.timeout = Settings.preneeds.timeout
+
+        conn.request :soap_headers
+
+        conn.response :preneeds_parser
+        conn.response :snakecase
+        conn.response :soap_to_json
```
Comment (saneshark) [inline lib/preneeds/configuration.rb]: going to change the name of this to something else, like xml_to_array_or_hash or xml_to_object

File: app/models/preneeds/branches_of_service.rb (hunk 20 above)
Comment (mphprogrammer) [inline app/models/preneeds/branches_of_service.rb]: I did, haven't pushed it up yet.

File: config/preneeds/xsd/preneeds.xsd (hunk 14 above)
Comment (mphprogrammer) [inline config/preneeds/xsd/preneeds.xsd]: Should be gone now.

File: lib/preneeds/models/attribute_types/xml_date.rb [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+module Preneeds
+  # This is a custom type class for ensuring Time is always coerced as UTC first then RFC1123
+  class XmlDate < Virtus::Attribute
+    def coerce(value)
+      return nil if value.to_s.empty?
+
+      begin
+        return Time.iso8601(value&.to_s).utc.strftime('%Y-%m-%d')
+      rescue ArgumentError
+        return value
```
Comment (saneshark) [inline lib/preneeds/models/attribute_types/xml_date.rb]: Not crazy about this rescue block, but I think it will probably be OK. Value would most likely be nil, I would think. Maybe that's OK. But doesn't hurt to show the error for debugging purposes maybe.

NITPICK.

**Comment by saneshark** [comment]
@patrickvinograd is there a vets.gov-team repo devops document with instructions for setting up a new host in dev/staging/production

I've stubbed the following to help instruct based on what Marc has done here in this PR. But could you review. If we prefer to use HOST or make other changes to be more consistent with other integrations we should probably consider that now rather than later.

https://github.com/department-of-veterans-affairs/vets.gov-team/blob/master/Products/Pre-Needs-Burial/api/devops/README.md

File: config/routes.rb (hunk 15 above)
Comment (mphprogrammer) [inline config/routes.rb]: @saneshark This is actually "namespaces" in the code (checked the preneeds-initial branch on github) - I am not sure why it is showing as scoped here...

**Comment by patrickvinograd** [comment]
@saneshark Please add a .md file under https://github.com/department-of-veterans-affairs/devops/tree/master/docs/External%20Service%20Integrations and follow one of the existing files there as an example of what we need - what you have here with hosts and relevant configuration is a good start; also include any relevant metrics, high level description of the nature of the data source (i.e. "it's a SOAP web service" is adequate), and most importantly POCs that somebody on call can use for support issues for both production and lower environments.
